README.md
###Smart-Form v2.x

Smart-Form is a form and list generator, which ensures a quick setup via simple array design.

## Basic Usage - Form

![Bildschirmfoto 2020-02-06 um 18 39 08](https://user-images.githubusercontent.com/10619091/73963402-7d607800-4910-11ea-932e-59a7496c67b1.png)

```php
<?php
include_once ("../include_form.php");

$arr ['form'] = array ('id' => 'form_newsletter','action' => 'ajax/handler.php','class' => 'segment attached','width' => '800','align' => 'center' );
$arr ['ajax'] = array ('success' => "$('#show_data').html(data);",'dataType' => 'html' );
$arr ['field'] ['date'] = array ('type' => 'date','label' => 'Date' );
$arr ['field'] ['firstname'] = array ('grid'=>'first', 'type' => 'input','label' => 'Firstname','placeholder' => 'Firstname' );
$arr ['field'] ['secondname'] = array ('grid'=>'second','type' => 'input','label' => 'Secondname','placeholder' => 'Secondname' );
$arr ['field'] ['grid'] = array ('type' => 'grid','class' => '','column' => [ "first" => '8',"second" => "8"] );
$arr ['field'] ['submit'] = array ('type' => 'button','value' => 'Submit','class' => 'submit','align' => 'center' );

$output_form = call_form ( $arr );
?>
<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8" />
<title>Formular - Small</title>
<link rel="stylesheet" href="../semantic/dist/semantic.min.css">
</head>
<body>
	<div class="ui main text container">
	<?=$output_form['html']?>
	<div id='show_data'></div>
	</div>
	<script src="../jquery/jquery.min.js"></script>
	<script src="../semantic/dist/semantic.min.js"></script>
	<?=$output_form['js']?>
</body>
</html>
```

## Basic Usage - List

![Bildschirmfoto 2020-02-06 um 18 38 58](https://user-images.githubusercontent.com/10619091/73963432-8d785780-4910-11ea-9039-a0e45c51cb4c.png)

```php
<?php
include ("../include_list.php");
$array = call_list ( 'inc/array_list.php', 'inc/mysql.php' );
?>

<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8" />
<title>Smart - List - Example</title>
<link rel="stylesheet" href="../semantic/dist/semantic.min.css">
</head>
<body>
	<div class="ui main text container">
	<?=$array['html']?>
	</div>
	<script src="../jquery/jquery.min.js"></script>
	<script src="../semantic/dist/semantic.min.js"></script>
	<script>var smart_form_wp = '../'</script>
	<script src="../js/smart_list.js"></script>
	<?=$array['js']?>
</body>
</html>
```

inc/array_list.php

```php
<?php
$arr ['list'] = array ('id' => 'demo_list','size' => 'small','class' => 'compact celled striped definitio' );

$arr ['mysql'] ['table'] = "list ";
$arr ['mysql'] ['field'] = "*";
$arr ['mysql'] ['like'] = 'firstname,secondname';
$arr ['mysql'] ['limit'] = '20';

$arr ['th'] ['firstname'] = array ('title' => "Firstname" );
$arr ['th'] ['secondname'] = array ('title' => "Secondname" );
$arr ['th'] ['birthday'] = array ('title' => "Birthday" );

$arr ['tr'] ['buttons'] ['left'] = array ('class' => 'tiny' );
$arr ['tr'] ['button'] ['left'] ['edit'] = array ('title' => '','icon' => 'edit','class' => 'blue mini','modal' => 'edit','popup' => 'Edit' );
$arr ['tr'] ['button'] ['left'] ['edit'] ['onclick'] = "$('#edit>.header').html('{firstname} {secondname}');";

$arr ['top'] ['buttons'] = array ('class' => 'tiny' );
$arr ['top'] ['button'] ['edit'] = array ('title' => 'Add new user','icon' => 'plus','class' => 'blue mini' );

$arr ['modal'] ['edit'] = array ('title' => 'Edit contact','url' => 'ajax/list_form_edit.php','class' => 'small' );
$arr ['modal'] ['edit'] ['button'] ['cancel'] = array ('title' => 'Close','color' => 'green','icon' => 'close' );
$arr ['modal'] ['edit'] ['button'] ['more'] = array ('title' => 'More','onclick' => "alert('test');" );

?>
```

inc/mysql.php

```php
<?php
$cfg_mysql['user']     = 'demo';
$cfg_mysql['password'] = 'demo12345';
$cfg_mysql['server']   = 'localhost';
$cfg_mysql['db']       = 'demo';

$GLOBALS['mysqli'] = new mysqli ( $cfg_mysql['server'], $cfg_mysql['user'], $cfg_mysql['password'], $cfg_mysql['db'] ) or die ( "Could not open connection to server {$cfg_mysql['server']}" );
?>
```

## License
[MIT](https://choosealicense.com/licenses/mit/)


CHANGELOG.md

### Changelog

## 2.3

> 30.11.2020

-add: form -> content -> scrolling
-add: list -> button top set href, target, popup 
-add: list -> modal -> close button hide with close_hide (view two delet-button if more then 10 rows)
-add: list -> auto reload toggle switsch on/off 
-mod: list -> order-dropdown now inner Filterline
-mod: list -> hide close x from modal
ALTER TABLE `list` ADD `category` VARCHAR(100) NOT NULL AFTER `message`; 


## 2.2

> 08.11.2020

-add: form -> line  'type' => 'line','text'=>'more'
-add: form -> content 'color' => 'red','size'=>'large', 'inverted' => true,
-add: list -> output for data array
-add: list -> set_tr_hover => 'red' or true  
-fix: list -> header colspan

## 2.1

> 28.09.2020

- add: form -> ajax -> xhr -> ''
- fix: form -> input -> coler
- add: list -> modal -> Buttons on the bottom of the modal possible (color,title,icon,onclick) form_id => ID from form to submit the form
- add: list -> change templates for popup title from modal Exmp. Title {title}
- add: list -> multiselect with checkboxes
- mod: new examples


## 2.0

> 12.02.2020 

- add: form -> label -> add class 'form ID'
- add: list -> total for int/double/....
- add: new structure for list 
- add: Redesign complete - new structure
- add: New examples 
- add: New documentation
- add: Clearable -> input/dropdown
- mod: $value = htmlspecialchars($value);
- mod: change usage - info in doc (develop)


## 1.x

> Version 1.9.6 (04.03.2019)

- fix: checkbox value 1 works now
- fix: checkbox onfocus works again

> Version 1.9.5 (19.02.2019)

- add: form -> search inner input
- fix: list -> reload -> use tooltip

> Version 1.9.4 (15.1.2019)

- fix: list -> export

> Version 1.9.3 (27.10.2018)

- add: form -> dropdown -> fulltext search

> Version 1.9.2 (01.06.2018)

- add: form -> dropdown -> get array from url

> Version 1.9.1 (11.05.2018)

- add: form -> radio -> default_select -> choose the first first in the radio
- add: list -> filter -> default_value -> Default value db

> Version 1.9 (26.02.2018)

- add: form -> select -> clear button
- add: list -> header='off' //Hide Header

> Version 1.8.2 (02.02.2017)

- add: form -> add_val_dropdown(key,value,name) //js add value inner Dropdown

> Version 1.8.1 (06.01.2017)

- add: form -> readonly for Textarea 
- fix: search-engine improved (faster load)

> Version 1.8.0 (08.12.2017)

- add: from -> autocomplete 
- add: form -> radio,checkbox -> onchange
- add: button -> tooltip
- add: all size now as well for Text "Checkbox inner Formular" (exmp.: class = 'ui small text' )
- add: list -> serial => true Show serial Nummber
- add: list -> Filter-select now colored (orange)
- add: list -> Modal -> url now with placeholder for ID {id}
- add: form -> input -> format -> %
- add: -> onchange -> you can use this placeholder for the {id}
- add: finder -> onchange -> you can use this placeholder for the {id}  `arr['finder']['onchange']`
- mod: input -> date -> select data-default-date()
- fix: correct date format for datepicker
- fix: jquery - plugin now on the right place

> Version 1.7.4 (04.11.2017)

- mod: input -> icon -> toolbar is hiding after use
- fix: input -> icon -> now icon with 'space usable'

> Version 1.7.2 (13.10.2017)

- fix: list -> global --> $GLOBALS for preg_replace_callback(...)

> Version 1.7.1 (29.09.2017)

- add: form -> type = 'icon'
- add: form -> label_class = 'admin' oder anderes

> Version 1.7 (30.08.2017)

- add: ckeditor v4.7.2
- add: form -> button -> text = value
- add: form -> label -> label = ' ' will be like 'label' => '&nbsp' for free space
- add: form -> button -> class_button = 'fluid red'
- add: form -> button -> tooltip = 'infotext für button'
- fix: form -> radio -> checked for default value works now

> Version 1.6.9 (14.08.2017)

- add: form -> checkbox -> onchange

> Version 1.6.8 16.07.2017

- add: color -> new > Version 
- mod: semantic 2.2.11

> Version 1.6.7 25.06.2017

- add: form -> label_right_id
- add: form -> input -> clear button "clear => true" 
- add: list -> id for buttons "top"

> Version 1.6.6. 28.05.2017

- fix: use form -> class after success 

> Version 1.6.5 15.04.2017

- add: form -> select -> array => array('title_1'=>Überschrift','1'=>'Eins','2' => 'Zwei')  //title_xxxx - für OptGroup

> Version 1.6.4 22.03.2017

- fix: show gallery direct after login (setting 'image_> Versions' => array ( 'thumbnail' => $thumbnail ) ); inner include_file_gallery.inc.php

> Version 1.6.3 20.03.2017

- add: form -> arr['value'] = array('value1'=>'1'); //Übergabe von Values

> Version 1.6.2 27.02.2017

- add: label_right for checkboxes
- mod: semantic-ui 2.2.8
- fix: dropdown call now inner HTMl for refresh

> Version 1.6.1 25.02.2017

- mod: auto-reload filter included

> Version 1.6 20.02.2017

- add: list -> auto_reload
- add: form -> Restore selectfields 

> Version 1.5 01.02.2017

- add: some bug fixes (close dialog and so one
- add: list -> tooltip for <th>
- add: list -> replace now with default option

> Version 1.4.8


include_form.php
<?php
/*
 * --------------------------------------------------------------------------------------------------------
 * | SMART - FORM (ssi-Product)
 * | Form-Generator: Using Semantic-UI Library
 * | 30.09.2020 - mm@ssi.at
 * | Version 2.1
 * --------------------------------------------------------------------------------------------------------
 */
error_reporting(E_ALL ^  E_NOTICE);
session_start ();
// Function call Formular
function call_form($arr) {
	date_default_timezone_set ( 'Europe/Vienna' );

	if (! is_array ( $arr )) {
		return "Form is not defined";
		exit ();
	}

	$dataType = "script";

	if ($arr ['form'] ['inline'])
		$inline = $arr ['form'] ['inline'];
	if ($arr ['ajax'] ['dataType'])
		$dataType = $arr ['ajax'] ['dataType'];

	// FORM - ID
	if ($arr ['form'] ['id']) {
		$form_id = $arr ['form'] ['id'];
		// Automatische Vergabe der ID
	} else {
		$form_id = 'smartForm' . ++ $GLOBALS ['ii'];
	}
	$data .= "\n\t\t form_id : '$form_id',"; // Wird auch übergeben und jquery-requests anpassen zu können (Bsp.: Messages)

	if ($arr ['sql'] ['query']) {

		$mysql_query = $GLOBALS ['mysqli']->query ( $arr ['sql'] ['query'] ) or die ( mysqli_error ( $GLOBALS ['mysqli'] ) );
		$mysql_array = mysqli_fetch_array ( $mysql_query );

		if ($arr ['sql'] ['key']) {
			$sql_key ['value'] = $mysql_array [$arr ['sql'] ['key']];
		} else
			// Wird automatisch genommen (erstes Feld)
			$sql_key ['value'] = $mysql_array [0];
	}

	// Default - Field ID
	$field_id = 100;

	if ($arr ['field']) {

		foreach ( $arr ['field'] as $key1 => $value1 ) {

			// Sollte keine ID vergeben worden sind wird eine erzeugt
			if (! $key1) {
				$field_id ++;
				$id = 'field-' . $field_id;
			} else if (! is_int ( $key1 ))
				$id = $key1;
			// Id darf keine Zahl sein, sonst wird diese erweitert durch field-
			else if (is_int ( $key1 )) {
				$id = 'field-' . $key1;
			}

			// ID - kann auch manuell gesetzt werden Bsp.: $arr['field']['domain']

			// if ($key1)

			foreach ( $value1 as $key2 => $value2 ) {
				${$key2} = $value2;
			}

			if ($read_only)
				$read_only = 'read-only';

			if ($disabled) {
				$disabled = "disabled='disabled' ";
			}

			// Jquery - Focus
			if ($focus) {
				if ($type == 'select' or $type == 'multiselect') {
					$jquery_focus = "$('#dropdown_$id').delay(800).focus(); ";
				} else
					$jquery_focus = "$('#$id').delay(800).focus(); ";

				$focus = '';
			}

			if ($arr ['value'] [$id])
				$value = $arr ['value'] [$id];

			if ($value === '0000-00-00')
				$value = '';
			if ($value_default && ! $value)
				$value = $value_default;

			$value_default = '';

			if (isset ( $mysql_array [$id] ) && ! $value) {
				// Auslesen der Werte aus der Datenbank
				// Wenn value nicht manuel gesetzt ist wird dieser aus der Datenbank genommen
				$value = $mysql_array [$id];
			}

			if (! is_array ( $value )) {
				// Macht Übergabe für Sonderzeichen in Inputfelder und Co möglich
				// $value = htmlspecialchars ( $value );
				$value = htmlspecialchars_decode ( $value );
			}

			// Infotext in einen popup
			if ($info) {
				// $info_tooltip = "<span data-tooltip='$info' ><i class='icon help circle tooltip grey'></i></span>";
				$info_tooltip = "<span title='$info' class='tooltip' ><i class='icon help circle grey'></i></span>";
			}

			if ($label === ' ')
				$label = '&nbsp;'; // Zeigt ein unsichbares Label an

			// Wird momentan in Dropdown eingesetzt
			if ($arr ['form'] ['size'] == 'massive')
				$option_style = 'font-size:22px';
			elseif ($arr ['form'] ['size'] == 'huge')
				$option_style = 'font-size:20px';
			elseif ($arr ['form'] ['size'] == 'big')
				$option_style = 'font-size:18px';
			elseif ($arr ['form'] ['size'] == 'large')
				$option_style = 'font-size:16px';
			elseif ($arr ['form'] ['size'] == 'small')
				$option_style = 'font-size:14px';
			elseif ($arr ['form'] ['size'] == 'tiny')
				$option_style = 'font-size:14px';
			elseif ($arr ['form'] ['size'] == 'mini')
				$option_style = 'font-size:12px';
			else
				$option_style = '';

			$onchange = preg_replace ( "/{id}/", $id, $onchange );

			if (file_exists ( __DIR__ . "/include/form/$type.php" ))
				include (__DIR__ . "/include/form/$type.php");
			elseif (in_array ( $type, array ('toggle','checkbox' ) ))
				// TOGGLE , CHECKBOX
				include (__DIR__ . "/include/form/checkbox.php");
			elseif ($type == 'explorer' or $type == 'finder')
				include_once (__DIR__ . "/include/form/finder.php");
			// INPUT,DATE, TIME
			elseif (in_array ( $type, array ('input','date','time' ) ))
				include (__DIR__ . "/include/form/input.php");
			// SELECT, MULTISELECT
			elseif ($type == 'dropdown' or $type == 'select' or $type == 'multiselect')
				include (__DIR__ . "/include/form/dropdown.php");

			/**
			 * **********************************************************************************************************
			 * END INPUTS and more
			 * **********************************************************************************************************
			 */

			if (is_array ( $rules ) or $validate) {
				$required = 'required';
			}

			// segment
			if ($segment) {
				if ($segment === true)
					$segment = '';
				$segment = "segment ui $segment";
				// message
			} elseif ($message) {
				if ($message === true)
					$message = '';
				$message = "message ui $message";
			}

			if (in_array ( $type, array ('hidden','tab','accordion','div','div_close','header' ) )) {
				$field .= $type_field;
			} else {

				/**
				 * *******************
				 * Field - Begin
				 * *******************
				 */
				//style='border:1px solid red;' 
				$field_div = "\n<div  id='row_$id' class='field row_field $required $class $segment $message'>";
				if ($label) {
					if ($label === true)
						$label = "&nbsp;";
					$field_div .= "<label class='$form_id label {$label_class}' id='label_$id'>$label $info_tooltip <span class='check_message'></span> $add_label_content</label>";
				}
				if (($type == 'input') and $label_right) { // Label im Input Right
					if ($label_right_click)
						$add_label_click = "onclick=\"$label_right_click\"";

					if ($label_right_tooltip)
						$add_label_right_tooltip = "data-tooltip='$label_right_tooltip'";

					if ($label_left_tooltip)
						$add_label_left_tooltip = " data-tooltip='$label_left_tooltip'";

					if (! $label_right_id)
						$label_right_id = "label_right_$id";

					$field_div .= "<div class='ui right labeled input'>$type_field<div id='$label_right_id' class='ui label $label_right_class' $add_label_click $add_label_right_tooltip>$label_right</div></div>";
				} else if (($type == 'input') and $label_left) { // Label im Input Left
					if ($label_left_click)
						$add_label_click = "onclick=\"$label_left_click\"";
					$field_div .= "<div class='ui left labeled input'><div id='label_left_$id' class='ui label $label_left_class' $add_label_click $add_label_left_tooltip>$label_left</div>$type_field</div>";
				} else if (($type == 'select' or $type == 'multiselect') and $label_right) { // Label im Input Right
					$field_div .= "<div class='ui right labeled input'>$type_field<div id='label_left_$id' class='ui label $label_right_class' $add_label_right_tooltip>$label_right</div></div>";
				} else if (($type == 'select' or $type == 'multiselect') and $label_left) { // Label im Input Left
					$field_div .= "<div class='ui labeled input'><div class='ui label'>$label_left</div>$type_field</div>";
				} else
					$field_div .= $type_field;

				$field_div .= $footer;
				$field_div .= "</div>";
			/**
			 * *******************
			 * Field - End
			 * *******************
			 */
			}

			if (! $grid) {
				$field .= $field_div;
			} else {

				//Übergabe an grid
				$grid_content [$grid] = $field_div;
			}

			$grid = '';
			$field_div = '';

			if ($type_field2) {
				$field .= "<div class='$required $class field'>";
				$field .= $type_field2;
				$field .= "</div>";
			}

			// Short - Version for valtiation
			if ($type == 'smart_password') {
				$valitation .= "'$id': { identifier: '$id', rules: [{ type: 'empty', prompt: 'Bitte Passwort eingeben'},{ type: 'length[8]', prompt: 'Das Passwort muss min. 8 Zeichen haben'},{ type:'containsNumbers', prompt: 'Passwort muss eine Zahl beinhalten' }] },";
				$valitation .= "'{$id}_repeat': { identifier: '{$id}_repeat', rules: [{ type: 'empty', prompt: 'Bitte Passwort eingeben'},{ type: 'match[$id]', prompt: 'Passwort stimmt nicht überein'}] },";
				$jquery .= "$.fn.form.settings.rules.containsNumbers = function(value){ var regex = new RegExp('[0-9]'); return regex.test(value); }";
			} else if ($validate == true) {
				if (in_array ( $type, array ('toggle','checkbox','slider' ) )) {
					$val_type = 'checked';
				} else {
					$val_type = 'empty';
				}

				if ($validate === true)
					$validate = 'empty';

				if (! $prompt)
					$prompt = 'Eingabe überprüfen';

				// Wenn mehr als 1 type gewählt wird, dann wird "empty" verwendet und
				if (str_word_count ( $validate ) > 1) {
					$prompt = "$validate";
					$validate = 'empty';
				}

				if ($id) {
					if ($validate == 'empty') {
						if ($type == 'radio')
							$val_type = 'checked';
						$valitation .= "'$id': { identifier: '$id', rules: [{ type: '$val_type', prompt: '$prompt'}] },";
					} elseif ($validate) {
						$valitation .= "'$id': { identifier: '$id', rules: [{ type: '$validate', prompt: '$prompt'},{ type: 'empty', prompt: '$prompt'}] },";
					}
				}
			}

			// Complex version with a lot of possibilties
			if (is_array ( $rules )) {
				$valitation_rules = json_encode ( $rules );
				$valitation .= "'$id': { rules: $valitation_rules },";
			} elseif ($rules) {
				$valitation .= "'$id': '$rules',";
			}

			if ($type == 'date') { // $('#$id').calendar('get date')
				$add_data .= "data.push({ name: '$id', value: change_calendar_data( $('#$id').calendar('get date') )  });";
			} else if ($type == 'radio') {
				// $data .= "\n\t\t '$id' : $('.$id:checked.$form_id').val(),";
				$add_data .= "data.push({ name: '$id', value: $('.$id:checked.$form_id').val() });";
			} else if ($type == 'recaptcha') {
				// Google captcha
				// echo 'g-recaptcha-response';
				// $data .= "\n\t\t 'recaptcha' : $('#g-recaptcha-response').val(),";
				$add_data .= "data.push({ name: 'recaptcha', value: $('#g-recaptcha-response').val() });";
			} else if ($type == 'ckeditor_inline') {
				// INLINE - CKEDITOR with DIV
				// $add .= "\n\t\t '$id' : $('#$id.$form_id').html(),";
			} elseif ($type == 'ckeditor5') {
				$add_data .= "data.push({ name: '$id', value: myEditor_$id.getData() });";
			} elseif ($id) {
				// Input, Textarea
				// $data .= "\n\t\t '$id' : $('#$id.$form_id').val(),";
			}

			// Tab
			if (is_array ( $arr ['tab'] ) and $tab) {
				$tab_field [$tab] .= $field;
				$tab = '';
				// Reguler
			} else {
				$set_field .= $field;
			}
			$field = '';

			$add_label_content = '';
			$data_html = "";
			$icon = "";
			$min = '';
			$max = '';
			$step = '';
			$toolbar = '';
			$message = '';
			$segment = '';
			$text = '';
			$size = '';
			$add_class = '';
			$id = '';
			$value = '';
			// unset($placeholder);
			$placeholder = '';
			$label = '';
			$type_field = '';
			$type = '';
			$required = '';
			$class = '';
			$rules = '';
			$validate = '';
			$valitation_rules = '';
			$date = '';
			$read_only = '';
			$prompt = '';
			$val_type = '';
			$type_field2 = '';
			$disabled = '';
			$class_input = '';
			$label_right = '';
			$label_left = '';
			$label_left_class = '';
			$label_left_click = '';
			$footer = '';
			$info_tooltip = '';
			$info = '';
			$onclick = '';
			$format = '';
			$options = '';
			$rows = '';
			$style = '';
			$onchange = '';
			$column = '';
		}
	} else {
		$field = "<h4 class='header ui'>Keine Felder definiert</h4>";
	}

	if ($arr ['button']) {
		// if (!is_array($arr['tab'])) $buttons = "<br>";
		$buttons .= "<div class='actions'><div class='field {$arr['buttons']['class']}' align='{$arr['buttons']['align']}' id='{$arr['buttons']['id']}'>";
		foreach ( $arr ['button'] as $key => $value ) {
			foreach ( $value as $key2 => $value2 ) {
				$$key2 = $value2;
			}

			if ($value) {
				if ($js)
					$set_onclick = "onclick=\"$js\" ";
				elseif ($onclick)
					$set_onclick = "onclick=\"$onclick\" ";

				if ($info)
					$add_class = 'tooltip';
				else
					$add_class = '';
				if ($type == 'submit' or $id == 'submit')
					$type = 'submit';
				else
					$type = '';

				if ($icon) {
					$icon = "<i class='icon $icon'></i> ";
					$class = "icon $class";
				}
				if ($tooltip)
					$tooltip = "data-tooltip='$tooltip'";
				$buttons .= "<div $tooltip class='ui $color $class button $form_id $key $add_class {$arr['form']['size']}' type='$type' id='$id' $set_onclick >$icon$value</div>";
				$id = '';
				$onclick = '';
				$type = '';
				$js = '';
				$tooltip = '';
				$icon = '';
			}
			$color = '';
			$class = '';
		}

		$buttons .= '</div></div>';
	}

	// print_r($arr_accordion_close);

	$hidden = '';
	// HIDDEN
	if ($arr ['hidden']) {
		foreach ( $arr ['hidden'] as $key => $value ) {
			$hidden .= "\n\t<input type=hidden name='$key' id='$key' value='$value' >";
			$data .= "\n\t\t $key : '$value' ,";
		}
	}
	
	// Hidden Field for db
	if ($sql_key ['value']) {
		$hidden .= "<input type=hidden name='update_id' id='update_id' value='{$sql_key['value']}' >";
		$data .= "\n\t\t update_id : '{$sql_key['value']}',";
	}

	// Default
	// $button_submit = 'Speichern';
	$button_reset = 'Abbrechen';
	$message_success = 'Speicherung war erfolgreich';

	if ($arr ['message'] ['success']) {
		$message_success = $arr ['message'] ['success'];
	}

	if ($arr ['button'] ['reset']) {
		$button_reset = $arr ['button'] ['reset'] ['value'];
	}

	// BREITE der Formulares
	if ($arr ['form'] ['width']) {
		$form_style = "max-width:{$arr['form']['width']}px;";
	}

	if ($arr ['header']) {
		if ($arr ['header'] ['icon']) {
			$add_header_icon = "<i class='{$arr['header']['icon']} icon'></i>";
			$add_header_icon_class = 'icon';
		}
		$output_form_open_header = "
		<div class='ui top $add_header_icon_class {$arr['header']['segment_class']}'>
			$add_header_icon
			<div class='content'>
				<div class='ui header {$arr['header']['class']}'>
					{$arr['header']['title']}
				</div>
			{$arr['header']['text']}
			</div>
		</div>";
	}

	if ($arr ['footer']) {
		$output_form_open_footer = "<div class='ui bottom {$arr['footer']['segment_class']}'>{$arr['footer']['text']}</div>";
	}

	// Form wird nur angezeigt wenn arr['form'] definiert worden ist
	if (is_array ( $arr ['form'] ) || is_array ( $arr ['ajax'] )) {
		if ($arr ['form'] ['align'])
			$output_form_open .= "<div align='{$arr['form']['align']}'>";

		$output_form_open .= "<div style='text-align:left; $form_style'>";
		$output_form_open .= $output_form_open_header;
		$output_form_open .= "<form id='$form_id' name ='$form_id' class='ui $segment_attached {$arr['form']['class']} {$arr['form']['size']} form'>";
		// wird benötigt wenn 'dataType' => html ist, wird bsp.: "ok" übergeben
		if (! $GLOBALS ['data_value']) {
			$output_form_close .= "<input id='data' name='data' type='hidden'>";
			$GLOBALS ['data_value'] = true;
		}
		$output_form_close .= "</form>";
		$output_form_close .= $output_form_open_footer;
		$output_form_close .= "</div>";

		if ($arr ['form'] ['align'])
			$output_form_close .= "</div>";
	}

	// if ($arr['header']) {
	// $output_form_close .= "</div></div>";
	// }

	/**
	 * ******* Auflistung Der Felder, wenn eingestellt mit TAB *****************
	 */

	$output_content .= $output_form_open;

	// Tab
	if (is_array ( $arr ['tab'] )) {

		// Darstellungsart TAB
		$tab_class = $arr ['tab'] ['class'];

		$content_class = $arr ['tab'] ['content_class'];
		if (! $tab_class) {
			$tab_class = "top attached";
			$content_class = "bottom attached $content_class";
		}

		foreach ( $arr ['tab'] ['tabs'] as $tab_key => $tab_value ) {
			if (! $arr ['tab'] ['active'])
				$arr ['tab'] ['active'] = 'first';
			if ($arr ['tab'] ['active'] == "$tab_key")
				$tab_active = 'active';
			else
				$tab_active = '';

			if ($content_class != 'basic') {
				$set_segment = 'segment ';
			}

			// Anzeigen wenn content vorhanden ist
			if ($tab_field [$tab_key]) {

				$tab_title .= "\n<a class='$tab_active item' id='$tab_key' data-tab='$tab_key'>$tab_value</a>";
				$tab_content .= "\n<div class='ui $set_segment tab $form_id {$arr['form']['size']} $content_class $tab_active' data-tab='$tab_key'>{$tab_field[$tab_key]}</div>";
			}
		}

		$output_content .= "<div id='tabgroup_$form_id'><div class='ui $tab_class tabular menu {$arr['form']['size']}'>$tab_title</div>$tab_content<div></div></div>";
		if ($content_class == 'basic') {
			$output_content .= "<br>";
		}

		if (! is_array ( $arr ['tab'] ))
			$output_content .= "<br>"; // Zeilenumbruch wenn TAB mit Einfassung verwendet wird (also keine spzifische Darstellung)

		$jquery .= "$('#tabgroup_$form_id .menu .item').tab();";
	}

	$output_content .= $set_field;

	if ($inline == 'list') {
		$output_content .= "<div style='text-align:left;' class='ui error message'></div>";
		$inline = '';
	}
	$output_content .= $buttons;
	$output_content .= $hidden;
	$output_content .= $output_form_close;

	/**
	 * **************************************************************************
	 */

	if (! $arr ['form'] ['keyboardShortcuts']) {
		$keyboardShowtcuts = "keyboardShortcuts : false,";
	} else
		$keyboardShowtcuts = "";

	$output ['html'] = $output_content;

	// Uebergibt POST/GET als json
	if ($arr ['ajax'] ['dataType'] == 'json') // json_stringify
		$json_stringify = 'JSON.stringify';

	// Standardmässig wird Form aufgerufen
	if ($arr ['form'] ['action'] !== false) {

		if (! $arr ['form'] ['action']) {
			$arr ['form'] ['action'] = $_SERVER ['PHP_SELF'];
		}

		if ($arr ['ajax'] ['xhr']) {
			$add_xhr = "
 			xhr: function(){
				var xhr = $.ajaxSettings.xhr();
				xhr.onprogress = function(e){
					data = e.currentTarget.responseText.substr(responseLen);
					responseLen = e.currentTarget.responseText.length;
					{$arr['ajax']['xhr']}
				};
				return xhr;
			},";
		}

		$add_function_submit .= "
						
		function submitForm_$form_id() {
		 var responseLen = 0;
		 var data = $('#$form_id').serializeArray();
		 $add_data
		 
		  $.ajax({
		    type: 'POST',
		    global   : false,
		    url: '{$arr['form']['action']}',
		    data: data,
		    dataType: '$dataType',
		    $add_xhr
		    beforeSend: function(){
		    	{$arr['ajax']['beforeSend']}
		    	$('#$form_id').attr('class','ui loading form $form_id {$arr['form']['size']} {$arr['form']['class']}');
			},
		    success: function(data){
		    	$('#$form_id').attr('class','ui form $form_id {$arr['form']['size']} {$arr['form']['class']}');
				$('.message .close').on('click', function() {
	  				$(this).closest('.message').fadeOut();
				});
				{$arr['ajax']['success']}
			}
		  });
		  return false;
		}
		";
	} // Wenn "'action' => false" wird nur success auf aufgerufen
	else {
		$add_function_submit .= "function submitForm_$form_id() {  {$arr['ajax']['success']} return false; }";
	}

	$output ['js'] .= "<script>";
	$output ['js'] .= "
					
		function change_calendar_data(date){
			if (date != null) {	
				date1 = new Date ( date );
				date = date1.getDate();
				year = date1.getFullYear();
				month = date1.getMonth()+1;
				return year+'-'+month+'-'+date;
			}
		}
		
		
		function get_icon(id,icon) {
			var icon = icon.replace('+', ' ');
			$('#button_icon_'+id).attr('class',icon+' large icon');
			$('#'+id).val(icon).change();
			$('.tooltip-click').popup('hide');
		}
		
	  	$(document).ready(function() {
	  	
			//$('.link.remove.icon').hide();
			//versteckt 'x' bei den Inputs wenn kein Eintrag vorhanden ist
			$('.ui-input').each(function () {
				if ($(this).val()) $('.link.remove.icon#icon_'+$(this).attr('id')).show();
				else  $('.link.remove.icon#icon_'+$(this).attr('id')).hide();
			});
			
			$('.ui-input').change( function(){
				if (!$(this).val())
					$('.link.remove.icon#icon_'+$(this).attr('id')).hide();
				else
					$('.link.remove.icon#icon_'+$(this).attr('id')).show();
			})
			
			
			$('.ui.accordion').accordion();
			$('.ui.accordion').accordion({ onOpen: function (item) {  $('.modal').modal('refresh'); } });
			//$('.tooltip').popup({ position:'right center' });
			$('.tooltip').popup();
			$('.tooltip-right').popup({ position:'right center' });
			$('.tooltip-top').popup({ position:'top center' });
			$('.tooltip-left').popup({ position:'left center' });
			$('.tooltip-click').popup({ on: 'click' });
			
			
	  		$jquery
	  		
		  	$('#$form_id.ui.form').form({
		  		fields: { $valitation },
		  		inline: '$inline',
		  		transition: 'fade down',
		  		$keyboardShowtcuts
		  		onSuccess: submitForm_$form_id
			});
			$jquery_focus
			{$arr['ajax']['onLoad']}
			{$arr['ajax']['onload']}
		});
		$add_function_submit";
	$output ['js'] .= "</script>";

	return $output;
}

include_once (__DIR__ . '/functions/filelist.php');